scalac-scoverage-runtime/shared/src/main/scala/scoverage/Invoker.scala

```diff
@@ -31,18 +35,27 @@ object Invoker {
     // times since for coverage we only care about 1 or more, (it just slows things down to
     // do it more than once), anything we can do to help is good. This helps especially with code
     // that is executed many times quickly, eg tight loops.
-    if (!ids.contains(dataDir, id)) {
+    if (!dataDirToIds.contains(dataDir)) {
+      // Guard against SI-7943: "TrieMap method getOrElseUpdate is not thread-safe".
+      dataDirToIds.synchronized {
+        if (!dataDirToIds.contains(dataDir)) {
+          dataDirToIds(dataDir) = ThreadSafeMap.empty[Int, Any]
+        }
+      }
+    }
+    val ids = dataDirToIds(dataDir)
+    if (!ids.contains(id)) {
       // Each thread writes to a separate measurement file, to reduce contention
       // and because file appends via FileWriter are not atomic on Windows.
       var files = threadFiles.get()
       if (files == null) {
-        files = ThreadSafeMap.empty[String, FileWriter]
+        files = mutable.HashMap.empty[String, FileWriter]
         threadFiles.set(files)
       }
       val writer = files.getOrElseUpdate(dataDir, new FileWriter(measurementFile(dataDir), true))
-      writer.append(id.toString + '\n').flush()
+      writer.append(Integer.toString(id)).append("\n").flush()
```

Comment:
I'm pretty sure that `id.toString` boxes here, so `Integer.toString(...)` saves that overhead.

The key change on this line is to replace the string concatenation with two separate `.append()` calls. The concatenation ends up implicitly constructing a `StringBuilder`.